File: doubly-linked-lists/exercises/dll-exercises.js
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
    this.prev = null;
  }
}

class DoublyLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = null;
  }
  traverse() {
    var currentNode = this.head;
    if (!this.head) {
      console.log("Empty list");
      return;
    }
    while (currentNode) {
      console.log(currentNode.value);
      currentNode = currentNode.next;
      if (currentNode == null) {
        break;
      }
    }
    console.log("Traversed");
  }
  push(val) {
    var newNode = new Node(val);
    if (!this.head) {
      this.head = newNode;
    } else {
      this.tail.next = newNode;
      newNode.prev = this.tail;
    }
    this.tail = newNode;
    this.length++;
    return this;
  }
  unshift(val) {
    var newNode = new Node(val);
    if (this.head === null) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.head.prev = newNode;
      newNode.next = this.head;
      this.head = newNode;
    }
    this.length++;
    return this;
  }
  shift() {
    var returnedNode = this.head;
    if (this.head === null) return undefined;
    if (this.length === 1) {
      this.head = 0;
      this.tail = 0;
    } else {
      this.head = returnedNode.next;
      this.head.prev = null;
      returnedNode.next = null;
    }
    this.length--;
    return returnedNode;
  }
  set(idx, val) {
    var currentNode = null;
    var counter;
    if (idx < 0 || idx >= this.length) return false;
    if (idx <= this.length / 2) {
      currentNode = this.head;
      counter = 0;
      while (counter != idx) {
        currentNode = currentNode.next;
        counter++;
      }
    } else {
      currentNode = this.tail;
      counter = this.length - 1;
      while (counter != idx) {
        currentNode = currentNode.prev;
        counter--;
      }
    }
    currentNode.value = val;
    return true;
  }
  remove(idx) {
    var counter;
    var currentNode;
    if (idx < 0 || idx >= this.length) return undefined;
    if (idx === this.length - 1) {
      currentNode = this.tail;
      currentNode.prev = this.tail;
      this.tail.next = null;
      currentNode.prev = null;
      currentNode.next = null;
      this.length--;
      return currentNode;
    }
    while (counter != idx) {
      currentNode = currentNode.next;
      counter++;
    }
    var previousNode = currentNode.prev;
    var afterNode = currentNode.next;
    previousNode.next = afterNode;
    afterNode.prev = previousNode;
    currentNode.prev = null;
    currentNode.next = null;
    this.length--;
    return currentNode;
  }
}

var l = new DoublyLinkedList();
l.push(0);
l.push(1);
l.push(2);
l.push(3);
l.set(0, 9);
l.traverse();
console.log(l.remove(0));
l.traverse();

File: doubly-linked-lists/exercises/dll-exercises.go
package main

import (
	"fmt"
)

type Node struct {
	value int
	prev  *Node
	next  *Node
}

type DLL struct {
	head   *Node
	tail   *Node
	length int
}

func (l *DLL) Traverse() {
	current := l.head
	if l.head == nil {
		fmt.Println("Empty list!")
		return
	}

	for {
		fmt.Println(current.value)
		current = current.next
		if current == nil {
			break
		}
	}
	fmt.Println("Traversed")
}

func (l *DLL) Push(n int) *DLL {
	newNode := &Node{value: n}
	if l.head == nil {
		l.head = newNode
	} else {
		l.tail.next = newNode
		newNode.prev = l.tail
	}
	l.tail = newNode
	l.length++
	return l
}

func (l *DLL) Unshift(n int) *DLL {
	newNode := &Node{value: n}
	if l.head == nil {
		l.head = newNode
		l.tail = newNode
	} else {
		l.head.prev = newNode
		newNode.next = l.head
		l.head = newNode
	}
	l.length++
	return l
}

func (l *DLL) Shift() *Node {
	shiftedNode := l.head
	if l.head == nil {
		return nil
	}
	if l.length == 1 {
		l.head = nil
		l.tail = nil
	} else {
		l.head = shiftedNode.next
		l.head.prev = nil
		shiftedNode.next = nil
	}
	l.length--
	return shiftedNode
}

func (l *DLL) Set(idx, val int) bool {
	var currentNode *Node
	var counter int

	if idx < 0 || idx >= l.length {
		return false
	}
	if idx <= int(l.length/2) {
		currentNode = l.head
		counter = 0
		for counter != idx {
			currentNode = currentNode.next
			counter++
		}
	} else {
		currentNode = l.tail
		counter = l.length - 1
		for counter != idx {
			currentNode = currentNode.prev
			counter--
		}
	}
	currentNode.value = val
	return true
}

func (l *DLL) Remove(idx int) *Node {
	removedNode := l.head
	counter := 0
	if idx < 0 || idx >= l.length {
		return nil
	}
	if idx == l.length-1 {
		removedNode = l.tail
		removedNode.prev = l.tail
		l.tail.next = nil
		removedNode.prev = nil
		removedNode.next = nil
		return removedNode
	}
	for counter != idx {
		removedNode = removedNode.next
		counter++
	}
	beforeNode := removedNode.prev
	afterNode := removedNode.next
	beforeNode.next = afterNode
	afterNode.prev = beforeNode
	removedNode.prev = nil
	removedNode.next = nil
	l.length--
	return removedNode
}

func main() {
	d := &DLL{}

	d.Push(0)
	d.Push(1)
	d.Push(2)
	d.Push(3)
	d.Remove(0)

	d.Traverse()
}
